File: swift/Sources/SconCore/Resolver.swift
import Foundation

final class Resolver {
    private final class EvalObject {
        var entries: [(String, EvalEntry)] = []
        var index: [String: Int] = [:]

        func get(_ key: String) -> EvalEntry? {
            index[key].map { entries[$0].1 }
        }

        func set(_ key: String, _ value: EvalEntry) {
            if let i = index[key] {
                entries[i] = (key, value)
            } else {
                index[key] = entries.count
                entries.append((key, value))
            }
        }
    }

    private final class EvalEntry {
        var value: Any
        var layer: String
        var kind: String

        init(_ value: Any, _ layer: String, _ kind: String) {
            self.value = value
            self.layer = layer
            self.kind = kind
        }
    }

    private var options: LoadOptions
    private var root = EvalObject()
    private var inProgress: [[String]] = [[]]
    var stack: [String] = []
    var seen: Set<String> = []
    private var cache: [String: Document] = [:]

    init(_ options: LoadOptions) {
        self.options = options
    }

    func eval(_ doc: Document) throws -> SconValue {
        try evalObject(doc.root, [], doc.file)
        return .object(publicObject(root))
    }

    private func evalObject(_ object: AstObject, _ path: [String], _ file: String?) throws {
        if path.count > options.maxObjectDepth {
            throw SconError(.ResourceLimitExceeded, "maximum object depth exceeded", span: object.span)
        }

        var localSeen = false
        for member in object.members {
            switch member {
            case .objectSpread(let spread):
                if localSeen {
                    throw SconError(.InvalidSpread, "object spread must appear before local members", span: spread.span)
                }
                guard let source = try lookup(spread.sub.path, spread.span).value as? EvalObject else {
                    throw SconError(.TypeMismatch, "object spread target is not an object", span: spread.span)
                }
                overlayBase(try objectAt(path, spread.span), source)
            case .include(let include):
                let included = try loadInclude(file, include)
                try evalObject(included.root, path, included.file)
            case .field(let field):
                localSeen = true
                try evalField(field, path, file)
            }
        }
    }

    private func evalField(_ field: AstField, _ current: [String], _ file: String?) throws {
        let target = current + field.path.segments.map(\.value)
        if case .object(let object, _) = field.value {
            try ensureObject(target, field.span)
            inProgress.append(target)
            defer { inProgress.removeLast() }
            try evalObject(object, target, file)
            return
        }
        try insert(target, evalValue(field.value, file), "ordinary", field.span)
    }

    private func evalValue(_ value: AstValue, _ file: String?) throws -> Any {
        switch value {
        case .null:
            return SconValue.null
        case .bool(let bool, _):
            return SconValue.bool(bool)
        case .number(let raw, let span):
            do {
                return SconValue.number(try SconNumber.parse(raw))
            } catch let error as SconError {
                throw SconError(error.code, error.message, span: span)
            }
        case .string(let string):
            return SconValue.string(try evalString(string))
        case .substitution(let sub):
            return cloneAny(try lookup(sub.path, sub.span).value)
        case .array(let array):
            return try evalArray(array, file)
        case .object(let object, _):
            let nested = Resolver(options)
            nested.stack = stack
            nested.seen = seen
            nested.cache = cache
            try nested.evalObject(object, [], file)
            return nested.root
        }
    }

    private func evalArray(_ array: AstArray, _ file: String?) throws -> SconValue {
        var out: [SconValue] = []
        for item in array.items {
            if out.count >= options.maxArrayLength {
                throw SconError(.ResourceLimitExceeded, "maximum array length exceeded", span: item.span)
            }
            switch item {
            case .value(let value, _):
                out.append(publicMaybe(try evalValue(value, file)))
            case .spread(let sub, let span):
                guard case .array(let values) = try lookup(sub.path, span).value as? SconValue else {
                    throw SconError(.TypeMismatch, "array spread target is not an array", span: span)
                }
                out.append(contentsOf: values.map(cloneValue))
            }
        }
        return .array(out)
    }

    private func evalString(_ string: AstString) throws -> String {
        if string.parts.count == 1, case .literal(let value) = string.parts[0] {
            return value
        }

        var out = ""
        for part in string.parts {
            switch part {
            case .literal(let value):
                out += value
            case .interpolation(let path, let span):
                let replacement = try lookup(path, span).value
                if case .string(let string) = replacement as? SconValue {
                    out += string
                } else if case .bool(let bool) = replacement as? SconValue {
                    out += bool ? "true" : "false"
                } else if case .number(let number) = replacement as? SconValue {
                    out += number.text
                } else {
                    throw SconError(.TypeMismatch, "interpolation requires string, number, or boolean", span: span)
                }
            }
        }
        return out
    }

    private func lookup(_ path: AstPath, _ span: Span) throws -> EvalEntry {
        let names = path.segments.map(\.value)
        if inProgress.contains(where: { $0 == names }) {
            throw SconError(.MissingReference, "reference is not completed yet", span: span)
        }

        var object = root
        var entry: EvalEntry?
        for (index, name) in names.enumerated() {
            entry = object.get(name)
            guard let current = entry else {
                throw SconError(.MissingReference, "missing reference '\(name)'", span: span)
            }
            if index < names.count - 1 {
                guard let next = current.value as? EvalObject else {
                    throw SconError(.TypeMismatch, "reference path crosses non-object value", span: span)
                }
                object = next
            }
        }
        return entry!
    }

    private func ensureObject(_ path: [String], _ span: Span) throws {
        var object = root
        for (index, name) in path.enumerated() {
            guard let entry = object.get(name) else {
                let child = EvalObject()
                object.set(name, EvalEntry(child, "local", "structural"))
                object = child
                continue
            }
            guard let next = entry.value as? EvalObject else {
                throw SconError(.PathConflict, "path conflicts with scalar value", span: span)
            }
            if index == path.count - 1 && entry.layer == "local" && entry.kind != "structural" {
                throw SconError(.PathConflict, "object field conflicts with ordinary value", span: span)
            }
            entry.layer = "local"
            entry.kind = "structural"
            object = next
        }
    }

    private func insert(_ path: [String], _ value: Any, _ kind: String, _ span: Span) throws {
        var object = root
        for name in path.dropLast() {
            if let entry = object.get(name) {
                guard let next = entry.value as? EvalObject else {
                    throw SconError(.PathConflict, "path conflicts with scalar value", span: span)
                }
                object = next
            } else {
                let child = EvalObject()
                object.set(name, EvalEntry(child, "local", "structural"))
                object = child
            }
        }

        let leaf = path.last!
        if let existing = object.get(leaf) {
            if existing.layer == "base" {
                overlayLocal(existing, value, kind)
            } else {
                throw SconError(.DuplicateKey, "duplicate key '\(leaf)'", span: span)
            }
        } else {
            object.set(leaf, EvalEntry(value, "local", kind))
        }
    }

    private func objectAt(_ path: [String], _ span: Span) throws -> EvalObject {
        var object = root
        for name in path {
            guard let entry = object.get(name) else {
                throw SconError(.PathConflict, "target object does not exist", span: span)
            }
            guard let next = entry.value as? EvalObject else {
                throw SconError(.PathConflict, "target path is not an object", span: span)
            }
            object = next
        }
        return object
    }

    private func loadInclude(_ file: String?, _ include: AstInclude) throws -> Document {
        let path = include.path.value
        if invalidIncludePath(path) {
            throw SconError(.InvalidIncludePath, "invalid include path", span: include.span)
        }

        let rootPath = URL(fileURLWithPath: options.includeRoot ?? (file.map { URL(fileURLWithPath: $0).deletingLastPathComponent().path } ?? ".")).standardizedFileURL.path
        let base = file.map { URL(fileURLWithPath: $0).deletingLastPathComponent().path } ?? rootPath
        let candidate = URL(fileURLWithPath: path, relativeTo: URL(fileURLWithPath: base)).standardizedFileURL.path
        if !candidate.hasPrefix(rootPath) {
            throw SconError(.IncludePathDenied, "include path escapes include root", span: include.span)
        }
        if stack.contains(candidate) {
            throw SconError(.IncludeCycle, "include cycle: \(candidate)", span: include.span)
        }
        if stack.count >= options.maxIncludeDepth {
            throw SconError(.ResourceLimitExceeded, "maximum include depth exceeded", span: include.span)
        }

        seen.insert(candidate)
        if seen.count > options.maxIncludeFiles {
            throw SconError(.ResourceLimitExceeded, "maximum include file count exceeded", span: include.span)
        }
        if let cached = cache[candidate] {
            return cached
        }

        var isDir: ObjCBool = false
        guard FileManager.default.fileExists(atPath: candidate, isDirectory: &isDir) else {
            throw SconError(.IncludeNotFound, "include file not found: \(candidate)", span: include.span)
        }
        if isDir.boolValue {
            throw SconError(.IncludeNotFile, "include path is not a file", span: include.span)
        }

        stack.append(candidate)
        defer { stack.removeLast() }
        do {
            let doc = try Parser.parseDocument(String(contentsOfFile: candidate, encoding: .utf8), file: candidate)
            cache[candidate] = doc
            return doc
        } catch let error as SconError {
            throw SconError(error.code == .InvalidRootType ? .IncludeRootTypeError : .IncludeParseError, error.message, span: error.span)
        }
    }

    private func invalidIncludePath(_ path: String) -> Bool {
        path.contains("://")
            || path.hasPrefix("classpath:")
            || path.contains("*")
            || path.hasPrefix("~")
            || path.hasPrefix("$")
            || path.hasPrefix("/")
            || path.range(of: #"^[A-Za-z]:[\\/]"#, options: .regularExpression) != nil
    }

    private func publicObject(_ object: EvalObject) -> SconObject {
        var out = SconObject()
        for (key, entry) in object.entries {
            out.set(key, publicMaybe(entry.value))
        }
        return out
    }

    private func publicMaybe(_ value: Any) -> SconValue {
        value is EvalObject ? .object(publicObject(value as! EvalObject)) : value as! SconValue
    }

    private func cloneAny(_ value: Any) -> Any {
        guard let object = value as? EvalObject else {
            return cloneValue(value as! SconValue)
        }

        let out = EvalObject()
        for (key, entry) in object.entries {
            out.set(key, EvalEntry(cloneAny(entry.value), entry.layer, entry.kind))
        }
        return out
    }

    private func cloneValue(_ value: SconValue) -> SconValue {
        switch value {
        case .array(let array):
            return .array(array.map(cloneValue))
        case .object(let object):
            var out = SconObject()
            for (key, value) in object.pairs {
                out.set(key, cloneValue(value))
            }
            return .object(out)
        default:
            return value
        }
    }

    private func overlayBase(_ target: EvalObject, _ source: EvalObject) {
        for (key, entry) in source.entries {
            if let existing = target.get(key) {
                if existing.layer == "base" {
                    overlayLocal(existing, cloneAny(entry.value), entry.kind)
                }
            } else {
                target.set(key, EvalEntry(cloneAny(entry.value), "base", "ordinary"))
            }
        }
    }

    private func overlayLocal(_ existing: EvalEntry, _ value: Any, _ kind: String) {
        if let target = existing.value as? EvalObject, let source = value as? EvalObject {
            mergeOverride(target, source)
            existing.layer = "local"
            existing.kind = kind
        } else {
            existing.value = value
            existing.layer = "local"
            existing.kind = kind
        }
    }

    private func mergeOverride(_ target: EvalObject, _ source: EvalObject) {
        for (key, entry) in source.entries {
            if let existing = target.get(key), let targetObject = existing.value as? EvalObject, let sourceObject = entry.value as? EvalObject {
                mergeOverride(targetObject, sourceObject)
            } else {
                target.set(key, EvalEntry(cloneAny(entry.value), entry.layer, entry.kind))
            }
        }
    }
}

File: csharp/src/Scon.Core/Resolver.cs
namespace RealmLabs.Scon;

internal sealed class Resolver
{
    private sealed class EvalObj { public readonly LinkedHashMap Entries = new(); }
    private sealed class EvalEntry(object value, string layer, string kind) { public object Value = value; public string Layer = layer; public string Kind = kind; }
    private sealed class LinkedHashMap : IEnumerable<KeyValuePair<string, EvalEntry>>
    {
        private readonly List<KeyValuePair<string, EvalEntry>> _entries = [];
        private readonly Dictionary<string, int> _index = new(StringComparer.Ordinal);
        public EvalEntry? Get(string key) => _index.TryGetValue(key, out var index) ? _entries[index].Value : null;
        public void Set(string key, EvalEntry value) { if (_index.TryGetValue(key, out var index)) _entries[index] = new(key, value); else { _index[key] = _entries.Count; _entries.Add(new(key, value)); } }
        public IEnumerator<KeyValuePair<string, EvalEntry>> GetEnumerator() => _entries.GetEnumerator(); IEnumerator IEnumerable.GetEnumerator() => GetEnumerator();
    }
    private readonly LoadOptions _options;
    private readonly EvalObj _root = new();
    private readonly List<List<string>> _inProgress = [[]];
    internal readonly List<string> Stack = [];
    internal readonly HashSet<string> Seen = new(StringComparer.Ordinal);
    private readonly Dictionary<string, Document> _cache = new(StringComparer.Ordinal);
    public Resolver(LoadOptions options) => _options = options;
    public SconValue Eval(Document doc) { EvaluateObject(doc.Root, [], doc.File); return PublicObject(_root); }
    private void EvaluateObject(AstObject obj, List<string> path, string? file)
    {
        if (path.Count > _options.MaxObjectDepth) throw new SconException(ErrorCode.ResourceLimitExceeded, "maximum object depth exceeded", obj.Span);
        var localSeen = false;
        foreach (var member in obj.Members)
        {
            switch (member)
            {
                case AstObjectSpread spread:
                    if (localSeen) throw new SconException(ErrorCode.InvalidSpread, "object spread must appear before local members", spread.Span);
                    if (Lookup(spread.Sub.Path, spread.Span).Value is not EvalObj source) throw new SconException(ErrorCode.TypeMismatch, "object spread target is not an object", spread.Span);
                    OverlayBase(ObjectAt(path, spread.Span), source);
                    break;
                case AstInclude include:
                    var included = LoadInclude(file, include); EvaluateObject(included.Root, path, included.File); break;
                case AstField field:
                    localSeen = true; EvalField(field, path, file); break;
            }
        }
    }
    private void EvalField(AstField field, List<string> current, string? file)
    {
        var target = current.Concat(field.Path.Segments.Select(s => s.Value)).ToList();
        if (field.Value is AstObjectValue obj) { EnsureObject(target, field.Span); _inProgress.Add(target); try { EvaluateObject(obj.Object, target, file); } finally { _inProgress.RemoveAt(_inProgress.Count - 1); } return; }
        Insert(target, EvalValue(field.Value, file), "ordinary", field.Span);
    }
    private object EvalValue(AstValue value, string? file) => value switch
    {
        AstNull => SconNull.Instance,
        AstBool b => new SconBool(b.Value),
        AstNumber n => ParseNumber(n),
        AstString s => new SconString(EvalString(s)),
        AstSubstitution s => CloneAny(Lookup(s.Path, s.Span).Value),
        AstArray a => EvalArray(a, file),
        AstObjectValue o => EvalNested(o, file),
        _ => throw new SconException(ErrorCode.UnexpectedToken, "unknown value"),
    };
    private static SconValue ParseNumber(AstNumber n) { try { return new SconNumberValue(SconNumber.Parse(n.Raw)); } catch (SconException ex) { throw new SconException(ex.Code, ex.Message, n.Span); } }
    private EvalObj EvalNested(AstObjectValue value, string? file) { var nested = new Resolver(_options); nested.Stack.AddRange(Stack); foreach (var s in Seen) nested.Seen.Add(s); foreach (var kv in _cache) nested._cache[kv.Key] = kv.Value; nested.EvaluateObject(value.Object, [], file); return nested._root; }
    private SconArray EvalArray(AstArray array, string? file)
    {
        var outArray = new SconArray();
        foreach (var item in array.Items)
        {
            if (outArray.Count >= _options.MaxArrayLength) throw new SconException(ErrorCode.ResourceLimitExceeded, "maximum array length exceeded", item.Span);
            if (item is AstArrayValue value) outArray.Add(PublicMaybe(EvalValue(value.Value, file)));
            else if (item is AstArraySpread spread) { if (Lookup(spread.Sub.Path, spread.Span).Value is not SconArray values) throw new SconException(ErrorCode.TypeMismatch, "array spread target is not an array", spread.Span); foreach (var v in values) outArray.Add(CloneValue(v)); }
        }
        return outArray;
    }
    private string EvalString(AstString value)
    {
        if (value.Parts.Count == 1 && value.Parts[0] is StringLiteral lit) return lit.Value;
        var outText = new StringBuilder();
        foreach (var part in value.Parts)
        {
            if (part is StringLiteral literal) outText.Append(literal.Value);
            else if (part is StringInterpolation interpolation)
            {
                var replacement = Lookup(interpolation.Path, interpolation.Span).Value;
                if (replacement is SconString s) outText.Append(s.Value);
                else if (replacement is SconBool b) outText.Append(b.Value ? "true" : "false");
                else if (replacement is SconNumberValue n) outText.Append(n.Number.ToSconString());
                else throw new SconException(ErrorCode.TypeMismatch, "interpolation requires string, number, or boolean", interpolation.Span);
            }
        }
        return outText.ToString();
    }
    private EvalEntry Lookup(AstPath path, Span span)
    {
        var names = path.Segments.Select(s => s.Value).ToList();
        if (_inProgress.Any(active => active.SequenceEqual(names))) throw new SconException(ErrorCode.MissingReference, "reference is not completed yet", span);
        var obj = _root; EvalEntry? entry = null;
        for (var i = 0; i < names.Count; i++) { entry = obj.Entries.Get(names[i]); if (entry is null) throw new SconException(ErrorCode.MissingReference, $"missing reference '{names[i]}'", span); if (i < names.Count - 1) { if (entry.Value is not EvalObj next) throw new SconException(ErrorCode.TypeMismatch, "reference path crosses non-object value", span); obj = next; } }
        return entry!;
    }
    private void EnsureObject(List<string> path, Span span)
    {
        var obj = _root;
        for (var i = 0; i < path.Count; i++)
        {
            var entry = obj.Entries.Get(path[i]);
            if (entry is null) { var child = new EvalObj(); obj.Entries.Set(path[i], new(child, "local", "structural")); obj = child; continue; }
            if (entry.Value is not EvalObj next) throw new SconException(ErrorCode.PathConflict, "path conflicts with scalar value", span);
            if (i == path.Count - 1 && entry.Layer == "local" && entry.Kind != "structural") throw new SconException(ErrorCode.PathConflict, "object field conflicts with ordinary value", span);
            entry.Layer = "local"; entry.Kind = "structural"; obj = next;
        }
    }
    private void Insert(List<string> path, object value, string kind, Span span)
    {
        var obj = _root;
        foreach (var name in path.Take(path.Count - 1))
        {
            var entry = obj.Entries.Get(name);
            if (entry is null) { var child = new EvalObj(); obj.Entries.Set(name, new(child, "local", "structural")); obj = child; }
            else { if (entry.Value is not EvalObj next) throw new SconException(ErrorCode.PathConflict, "path conflicts with scalar value", span); obj = next; }
        }
        var leaf = path[^1]; var existing = obj.Entries.Get(leaf);
        if (existing is null) obj.Entries.Set(leaf, new(value, "local", kind));
        else if (existing.Layer == "base") OverlayLocal(existing, value, kind);
        else throw new SconException(ErrorCode.DuplicateKey, $"duplicate key '{leaf}'", span);
    }
    private EvalObj ObjectAt(List<string> path, Span span)
    {
        var obj = _root;
        foreach (var name in path) { var entry = obj.Entries.Get(name); if (entry is null) throw new SconException(ErrorCode.PathConflict, "target object does not exist", span); if (entry.Value is not EvalObj next) throw new SconException(ErrorCode.PathConflict, "target path is not an object", span); obj = next; }
        return obj;
    }
    private Document LoadInclude(string? file, AstInclude include)
    {
        var path = include.Path.Value;
        if (InvalidIncludePath(path)) throw new SconException(ErrorCode.InvalidIncludePath, "invalid include path", include.Span);
        var rootPath = Path.GetFullPath(_options.IncludeRoot ?? (file is null ? "." : Path.GetDirectoryName(file)!));
        var basePath = file is null ? rootPath : Path.GetDirectoryName(file)!;
        var candidate = Path.GetFullPath(Path.Combine(basePath, path));
        if (!candidate.StartsWith(rootPath, StringComparison.Ordinal)) throw new SconException(ErrorCode.IncludePathDenied, "include path escapes include root", include.Span);
        if (Stack.Contains(candidate)) throw new SconException(ErrorCode.IncludeCycle, $"include cycle: {candidate}", include.Span);
        if (Stack.Count >= _options.MaxIncludeDepth) throw new SconException(ErrorCode.ResourceLimitExceeded, "maximum include depth exceeded", include.Span);
        Seen.Add(candidate); if (Seen.Count > _options.MaxIncludeFiles) throw new SconException(ErrorCode.ResourceLimitExceeded, "maximum include file count exceeded", include.Span);
        if (_cache.TryGetValue(candidate, out var cached)) return cached;
        if (!File.Exists(candidate)) throw new SconException(ErrorCode.IncludeNotFound, $"include file not found: {candidate}", include.Span);
        if ((File.GetAttributes(candidate) & FileAttributes.Directory) != 0) throw new SconException(ErrorCode.IncludeNotFile, "include path is not a file", include.Span);
        if (new FileInfo(candidate).Length > _options.MaxFileSize) throw new SconException(ErrorCode.ResourceLimitExceeded, "maximum file size exceeded", include.Span);
        Stack.Add(candidate);
        try { var doc = Parser.ParseDocument(File.ReadAllText(candidate), candidate); _cache[candidate] = doc; return doc; }
        catch (SconException ex) { throw new SconException(ex.Code == ErrorCode.InvalidRootType ? ErrorCode.IncludeRootTypeError : ErrorCode.IncludeParseError, ex.Message, ex.Span); }
        finally { Stack.RemoveAt(Stack.Count - 1); }
    }
    private static bool InvalidIncludePath(string path) => path.Contains("://", StringComparison.Ordinal) || path.StartsWith("classpath:", StringComparison.Ordinal) || path.Contains('*') || path.StartsWith('~') || path.StartsWith('$') || Path.IsPathRooted(path) || Regex.IsMatch(path, "^[A-Za-z]:[\\\\/]");
    private static SconObject PublicObject(EvalObj obj) { var outObj = new SconObject(); foreach (var (key, entry) in obj.Entries) outObj.Set(key, PublicMaybe(entry.Value)); return outObj; }
    private static SconValue PublicMaybe(object value) => value is EvalObj obj ? PublicObject(obj) : (SconValue)value;
    private static object CloneAny(object value) { if (value is not EvalObj obj) return CloneValue((SconValue)value); var outObj = new EvalObj(); foreach (var (key, entry) in obj.Entries) outObj.Entries.Set(key, new(CloneAny(entry.Value), entry.Layer, entry.Kind)); return outObj; }
    private static SconValue CloneValue(SconValue value) => value switch { SconArray a => new SconArray(a.Select(CloneValue)), SconObject o => CloneObject(o), _ => value };
    private static SconObject CloneObject(SconObject obj) { var outObj = new SconObject(); foreach (var (key, value) in obj) outObj.Set(key, CloneValue(value)); return outObj; }
    private static void OverlayBase(EvalObj target, EvalObj source) { foreach (var (key, entry) in source.Entries) { var existing = target.Entries.Get(key); if (existing is null) target.Entries.Set(key, new(CloneAny(entry.Value), "base", "ordinary")); else if (existing.Layer == "base") OverlayLocal(existing, CloneAny(entry.Value), entry.Kind); } }
    private static void OverlayLocal(EvalEntry existing, object value, string kind) { if (existing.Value is EvalObj a && value is EvalObj b) { MergeOverride(a, b); existing.Layer = "local"; existing.Kind = kind; } else { existing.Value = value; existing.Layer = "local"; existing.Kind = kind; } }
    private static void MergeOverride(EvalObj target, EvalObj source) { foreach (var (key, entry) in source.Entries) { var existing = target.Entries.Get(key); if (existing?.Value is EvalObj a && entry.Value is EvalObj b) MergeOverride(a, b); else target.Entries.Set(key, new(CloneAny(entry.Value), entry.Layer, entry.Kind)); } }
}
